File: Data Structures & Algorithms/minimum-size-subarray-sum/submission-7.py
class Solution:
    def minSubArrayLen(self, target: int, nums: List[int]) -> int:
        l, total, length = 0, 0, float('inf')
        for r in range(len(nums)):
            total += nums[r]
            while total >= target:
                length = min(r - l + 1, length)
                total -= nums[l]
                l += 1
        return 0 if length == float('inf') else length

File: Data Structures & Algorithms/partition-equal-subset-sum/submission-7.rs
impl Solution {
    pub fn can_partition(nums: Vec<i32>) -> bool {
        let total: usize = nums.iter().map(|&num| num as usize).sum();

        if !total.is_multiple_of(2) {
            return false;
        }

        let target = total / 2;
        let mut dp = vec![false; target + 1];
        dp[0] = true;

        for num in nums {
            let value = num as usize;

            for cur_sum in (value..=target).rev() {
                dp[cur_sum] = dp[cur_sum] || dp[cur_sum - value]
            }
        }

        dp[target]
    }
}
